File: sdk/python/omni_connector/connector.py
from __future__ import annotations

import logging
from abc import ABC, abstractmethod
from typing import TYPE_CHECKING, Any

from .context import SyncContext
from .models import ActionDefinition, ActionResponse, ConnectorManifest, SearchOperator

if TYPE_CHECKING:
    from mcp.client.stdio import StdioServerParameters

    from .mcp_adapter import McpAdapter

logger = logging.getLogger(__name__)


class Connector(ABC):
    """Base class for Omni connectors."""

    def __init__(self) -> None:
        self._cancelled_syncs: set[str] = set()
        self._mcp_adapter: McpAdapter | None = None

    @property
    @abstractmethod
    def name(self) -> str:
        """Connector name (e.g., 'google-drive', 'slack')."""
        pass

    @property
    @abstractmethod
    def version(self) -> str:
        """Connector version (semver)."""
        pass

    @property
    @abstractmethod
    def source_types(self) -> list[str]:
        """Source type slugs this connector handles (e.g., ['google_drive', 'gmail'])."""
        pass

    @property
    def display_name(self) -> str:
        """Human-readable display name. Override to customize."""
        return self.name

    @property
    def description(self) -> str:
        """Short description for the UI. Override to customize."""
        return ""

    @property
    def sync_modes(self) -> list[str]:
        """Supported sync modes. Override to customize."""
        return ["full"]

    @property
    def actions(self) -> list[ActionDefinition]:
        """Available connector actions. Override to add actions."""
        return []

    @property
    def search_operators(self) -> list[SearchOperator]:
        """Search operators this connector supports. Override to declare operators."""
        return []

    @property
    def mcp_command(self) -> StdioServerParameters | None:
        """Return stdio params for an external MCP server binary.

        Override this property to enable MCP support. The SDK will spawn the
        server as a subprocess, communicate via stdio transport, and expose
        its tools, resources, and prompts through the Omni protocol.

        Example::

            @property
            def mcp_command(self):
                return StdioServerParameters(
                    command="github-mcp-server",
                    args=["stdio"],
                )
        """
        return None

    @property
    def mcp_adapter(self) -> McpAdapter | None:
        if self._mcp_adapter is not None:
            return self._mcp_adapter
        params = self.mcp_command
        if params is None:
            return None
        from .mcp_adapter import McpAdapter

        self._mcp_adapter = McpAdapter(params)
        return self._mcp_adapter

    async def bootstrap_mcp(self, credentials: dict[str, Any]) -> None:
        """Discover MCP tools/resources/prompts and cache them.

        Called when credentials first become available (e.g., during initial sync).
        Spawns a temporary subprocess, introspects it, caches the results, then
        shuts down. Subsequent manifest builds use the cache.
        """
        adapter = self.mcp_adapter
        if adapter is None:
            logger.debug("bootstrap_mcp: no MCP adapter, skipping")
            return
        env = self.prepare_mcp_env(credentials)
        logger.info(
            "Bootstrapping MCP: discovering tools (env keys: %s)",
            sorted(env.keys()) if env else [],
        )
        try:
            await adapter.discover(env)
        except Exception:
            logger.warning("MCP bootstrap failed", exc_info=True)

    async def _get_all_actions(self) -> list[ActionDefinition]:
        """Merge manually-defined actions with MCP-derived actions."""
        manual_actions = self.actions
        adapter = self.mcp_adapter
        if adapter is None:
            return manual_actions
        try:
            mcp_actions = await adapter.get_action_definitions()
        except Exception:
            logger.warning("Failed to list MCP tools", exc_info=True)
            return manual_actions
        manual_names = {a.name for a in manual_actions}
        merged = list(manual_actions)
        for action in mcp_actions:
            if action.name not in manual_names:
                merged.append(action)
        return merged

    async def get_manifest(self, connector_url: str) -> ConnectorManifest:
        """Return connector manifest."""
        adapter = self.mcp_adapter
        resources = []
        prompts = []
        if adapter is not None:
            try:
                resources = await adapter.get_resource_definitions()
            except Exception:
                logger.warning("Failed to list MCP resources", exc_info=True)
            try:
                prompts = await adapter.get_prompt_definitions()
            except Exception:
                logger.warning("Failed to list MCP prompts", exc_info=True)
        return ConnectorManifest(
            name=self.name,
            display_name=self.display_name,
            version=self.version,
            sync_modes=self.sync_modes,
            connector_id=self.name,
            connector_url=connector_url,
            source_types=self.source_types,
            description=self.description,
            actions=await self._get_all_actions(),
            search_operators=self.search_operators,
            mcp_enabled=adapter is not None,
            resources=resources,
            prompts=prompts,
        )

    @abstractmethod
    async def sync(
        self,
        source_config: dict[str, Any],
        credentials: dict[str, Any],
        state: dict[str, Any] | None,
        ctx: SyncContext,
    ) -> None:
        """
        Execute a sync operation.

        Args:
            source_config: Source configuration from database
            credentials: Authentication credentials
            state: Previous sync state for incremental syncs
            ctx: Sync context with emit(), complete(), etc.
        """
        pass

    def cancel(self, sync_run_id: str) -> bool:
        """
        Handle cancellation request.

        Returns True if sync was found and marked for cancellation.
        """
        self._cancelled_syncs.add(sync_run_id)
        return True

    def prepare_mcp_env(self, credentials: dict[str, Any]) -> dict[str, str]:
        """Return env vars for the MCP subprocess given Omni credentials.

        Override this to bridge Omni credentials to the env vars your MCP
        server expects. The returned dict is merged into the subprocess env.

        Example::

            def prepare_mcp_env(self, credentials):
                return {"GITHUB_PERSONAL_ACCESS_TOKEN": credentials["token"]}
        """
        return {}

    async def execute_action(
        self,
        action: str,
        params: dict[str, Any],
        credentials: dict[str, Any],
    ) -> ActionResponse:
        """
        Execute a connector action.

        Override this method to implement connector-specific actions.
        If MCP is enabled and the action matches an MCP tool, it is
        dispatched to the MCP server automatically.
        """
        adapter = self.mcp_adapter
        if adapter is not None:
            env = self.prepare_mcp_env(credentials)
            mcp_tool_names = {a.name for a in await adapter.get_action_definitions(env)}
            if action in mcp_tool_names:
                return await adapter.execute_tool(action, params, env)
        return ActionResponse.not_supported(action)

    def serve(self, port: int = 8000, host: str = "0.0.0.0") -> None:
        """Start the HTTP server for this connector."""
        import uvicorn

        from .server import create_app

        app = create_app(self)
        logger.info("Starting %s connector on %s:%d", self.name, host, port)
        uvicorn.run(app, host=host, port=port)

File: sdk/typescript/src/connector.ts
import type { SyncContext } from './context.js';
import type {
  ConnectorManifest,
  ActionDefinition,
  ActionResponse,
  SearchOperator,
} from './models.js';
import { createActionResponseNotSupported } from './models.js';
import { createServer } from './server.js';
import { getLogger } from './logger.js';

export interface ServeOptions {
  port?: number;
  host?: string;
}

export abstract class Connector<
  TConfig extends Record<string, unknown> = Record<string, unknown>,
  TCredentials extends Record<string, unknown> = Record<string, unknown>,
  TState extends Record<string, unknown> = Record<string, unknown>,
> {
  abstract readonly name: string;
  abstract readonly version: string;
  abstract readonly sourceTypes: string[];

  private _mcpAdapter: unknown | null = null;

  get displayName(): string {
    return this.name;
  }

  get description(): string {
    return '';
  }

  readonly syncModes: string[] = ['full'];
  readonly actions: ActionDefinition[] = [];
  readonly searchOperators: SearchOperator[] = [];
  readonly extraSchema?: Record<string, unknown>;
  readonly attributesSchema?: Record<string, unknown>;

  /**
   * Return an MCP McpServer instance if this connector supports MCP.
   * Override this getter to enable MCP support.
   * Requires @modelcontextprotocol/sdk as a dependency.
   */
  get mcpServer(): unknown | undefined {
    return undefined;
  }

  async getMcpAdapter(): Promise<{ getActionDefinitions(): Promise<ActionDefinition[]>; getResourceDefinitions(): Promise<unknown[]>; getPromptDefinitions(): Promise<unknown[]>; executeTool(name: string, params: Record<string, unknown>): Promise<ActionResponse>; readResource(uri: string): Promise<unknown>; getPrompt(name: string, args?: Record<string, string>): Promise<unknown> } | undefined> {
    if (this._mcpAdapter !== null) {
      return this._mcpAdapter as ReturnType<typeof this.getMcpAdapter> extends Promise<infer T> ? T : never;
    }
    const server = this.mcpServer;
    if (!server) {
      return undefined;
    }
    const { McpAdapter } = await import('./mcp-adapter.js');
    this._mcpAdapter = new McpAdapter(server as any);
    return this._mcpAdapter as any;
  }

  private async getAllActions(): Promise<ActionDefinition[]> {
    const manualActions = this.actions;
    const adapter = await this.getMcpAdapter();
    if (!adapter) {
      return manualActions;
    }
    const mcpActions = await adapter.getActionDefinitions();
    const manualNames = new Set(manualActions.map((a) => a.name));
    return [...manualActions, ...mcpActions.filter((a) => !manualNames.has(a.name))];
  }

  async getManifest(connectorUrl: string): Promise<ConnectorManifest> {
    const adapter = await this.getMcpAdapter();
    return {
      name: this.name,
      display_name: this.displayName,
      version: this.version,
      sync_modes: this.syncModes,
      connector_id: this.name,
      connector_url: connectorUrl,
      source_types: this.sourceTypes,
      description: this.description,
      actions: await this.getAllActions(),
      search_operators: this.searchOperators,
      extra_schema: this.extraSchema,
      attributes_schema: this.attributesSchema,
      mcp_enabled: adapter !== undefined,
      resources: adapter ? await adapter.getResourceDefinitions() as any[] : [],
      prompts: adapter ? await adapter.getPromptDefinitions() as any[] : [],
    };
  }

  abstract sync(
    sourceConfig: TConfig,
    credentials: TCredentials,
    state: TState | null,
    ctx: SyncContext
  ): Promise<void>;

  cancel(_syncRunId: string): boolean {
    return false;
  }

  /**
   * Set up environment for MCP tool/resource/prompt calls.
   * Override to bridge Omni credentials to the env vars your MCP server expects.
   */
  prepareMcpEnv(_credentials: TCredentials): void {
    // no-op by default
  }

  async executeAction(
    action: string,
    params: Record<string, unknown>,
    credentials: TCredentials
  ): Promise<ActionResponse> {
    const adapter = await this.getMcpAdapter();
    if (adapter) {
      const mcpActions = await adapter.getActionDefinitions();
      const mcpToolNames = new Set(mcpActions.map((a) => a.name));
      if (mcpToolNames.has(action)) {
        this.prepareMcpEnv(credentials);
        return adapter.executeTool(action, params);
      }
    }
    return createActionResponseNotSupported(action);
  }

  serve(options: ServeOptions = {}): void {
    const port = options.port ?? parseInt(process.env.PORT ?? '8000', 10);
    const host = options.host ?? '0.0.0.0';

    const app = createServer(this);
    const logger = getLogger(this.name);
    app.listen(port, host, () => {
      logger.info(`Connector ${this.name} v${this.version} listening on ${host}:${port}`);
    });
  }
}
